### ISSUE TITLE
Immersive Enginnering has glitches with Optifine

### ISSUE THREAD

**Issue by LuxerTheGamer** [sp614x/optifine#1024]
Hi! I found two bugs, when I was playing on immersive enginnering with optifine. On immersive enginnering github page I saw this problem, but the immersive enginnering creator said, this is problem on the optifine side. Probably this is true.

The first problem is a incorrect texture on crusher (multiblock construct), probably texture is enlarged.
Crusher with optifine:
https://imgur.com/a/BtiYx
Crusher without optifine:
https://imgur.com/a/2VAG8

The second problem is with revolver. Probably, the element of the weapon with optifine doesn't have a texture, or texture is transparent.
Revolver with optifine:
https://imgur.com/a/7e8Uf
Revolver without optifine:
https://imgur.com/a/tvw7n

Thanks for help, and sorry for my english.

**Comment by malte0811** [comment]
(I'm a developer/collaborator at/of/... Immersive Engineering)
The first one (crusher) is probably OF's fault, I don't think there is anything weird about the crusher model, if there is, please tell me/us.
The second one (revolver) is most likely IE's fault and we don't expect you to support/fix it. @LuxerTheGamer Try disabling "fancyItemAnimations" in the config, that will most likely fix it. @sp614x Blu implemented some weird magic self-modifying baked quads to make item animations possible without a coremod (see BluSunrize/ImmersiveEngineering@06fb9744c3097e3d096c1694c4e3c3b5c3522906). We know doing this can cause all sorts of bugs (see the commit note), harms FPS (even though not as badly as one may think), and we do not expect it to ever work with OF, so you can probably ignore it.

**Comment by sp614x** [comment]
@malte0811 dynamic baked quads are bad, the vertex data is expected to be initialized once and never changed. Dynamic models which use a set of predefined quads are OK. 

@LuxerTheGamer have you tried disabling shaders? How should the crusher look like?

**Comment by malte0811** [comment]
I'm not quite sure what you are talking about in the first half, but that may be because I never tried to make dynamic models myself. How would we implement dynamic/moving models without modifying the vertex data?
The crusher model looks as it should, the textures look like they are zoomed in (one pixel isn't 1/16 blocks any more).

**Comment by sp614x** [comment]
The BakedQuad is expected to have static vertex data. It should be initialized once and then is should stay the same. On the other hand the `IBakedModel.getQuads()` is free to return whatever list of quads it wants.

Are the crusher textures having power-of-2 dimensions? 
With mipmaps enabled all model textures should have power-of-2 dimensions.

**Comment by malte0811** [comment]
I'll have to check how exactly the quads work. The crusher textures are 256x256. There are a couple of textures that aren't powers of 2, I'll try to change them before the next release.

**Comment by LuxerTheGamer** [comment]
@malte0811 
First I tried disable "fancyItemAnimations" and It changed nothing.
@sp614x 
I tried disable the shaders first, also It changed nothing. Crusher should look like on this photo: https://imgur.com/a/2VAG8

**Comment by malte0811** [comment]
What version of IE (and Optifine) are you using? There was a bug before IE build 73 where the revolvers would look like this when the animations are disabled. OF might have a similar effect on the models?

**Comment by LuxerTheGamer** [comment]
@malte0811 
Version: 1.12.2
Mods:
ImmersiveEngineering-0.12-75 and older, on 1.12.2
Forge 14.23.0.2559 and older, on 1.12.2
Optifine C5/C6/C7 1.12.2

**Comment by sp614x** [comment]
The crusher textures look like they are zoomed 2x, very strange.
OptiFine is not changing anything related to handlePerspective().

**Comment by LuxerTheGamer** [comment]
@sp614x Yes it's look like the textures are zoomed probably 2x. I found the crusher texture in _ImmersiveEngineering-0.12-75.jar/assests/immersiveengineering/textures/models/crusher.png_. The 
weird thing is when I tried to changed crusher textures, nothing was changed with optifine and without optifine. It's a joke? Where is a real localization this textures? I also tried add purebdcraft immersive enginnering addon 1.7.10 to purebdcraft resourcepack 1.12.2 and nothing was changed in immersive enginnering.

**Comment by LuxerTheGamer** [comment]
@malte0811 Where is crusher real localization in .jar?

**Comment by malte0811** [comment]
That is supposed to be the real texture AFAIK. Are you sure your launcher isn't overriding the changed jar?

**Comment by LuxerTheGamer** [comment]
@malte0811 Yes I'm sure because I use orginal minecraft launcher from Mojang

**Comment by Dockter** [comment]
I would like to add that we had a similar issue and our issue was this:  https://github.com/AlmuraDev/Almura/blob/bleeding/src/main/java/com/almuradev/content/model/obj/OBJBakedModel.java#L113

Commenting that line out of our loader fixed our issue.

**Comment by malte0811** [comment]
@Dockter Thank you, that helps a LOT!

@sp614x This is somewhere between OF and Forge: [Here](https://github.com/MinecraftForge/MinecraftForge/blob/d0bdb233000ac9aff049c5ae2ad8043480a96af9/src/main/java/net/minecraftforge/client/model/pipeline/UnpackedBakedQuad.java#L183-L186) is the (really) relevant bit of code. The crusher textures ends up right in the corner of the texture map (0,0, possibly because it is very large?). Without OF, both `getOrigin{X/Y}` and `getMin{U/V}` return exactly 0. With OF, `getOrigin` returns 0, but `getMin` returns something like 1e-6. In both cases `tX`, `tY`, `tS` and `ep` are set to unusual values. The main diffference is that with OF `t*` ends up as +0, causing `ep` to become `+\infty`. This causes a large shift of the UV's [here](https://github.com/MinecraftForge/MinecraftForge/blob/d0bdb233000ac9aff049c5ae2ad8043480a96af9/src/main/java/net/minecraftforge/client/model/pipeline/UnpackedBakedQuad.java#L224). Without OF all values end up as `NaN` since the first line is 0/0. All later checks fail and the shift remains very small.
Somewhat related question: Is there a reason there aren't any versions of OF that work in a deobf environment? I know there are ways to make it work, but it would be nice to not have to fiddle with everything for 2 hours before they work. Being able to place breakpoints and change code easily helps a lot with debugging.

**Comment by sp614x** [comment]
OptiFine does not have a version compatible with Forge deobf environment.
The only way to do it is via deobfuscation: https://github.com/octarine-noise/simpledeobf

**Comment by LuxerTheGamer** [comment]
@sp614x Will it be fixed in future version of Optifine?
Because using the simpledeobf is too hard for me ...

**Comment by sp614x** [comment]
There are no plans to make OF compatible with Forge deobf environment, mostly because it is too complex.

**Comment by malte0811** [comment]
I'm not sure whether that question was about compatibility with forge dev envs. If it was about the bug you reported: I'll try to make a pull request with a fix in forge today. The algorithm is broken without OF, it just breaks in a non-visible way.

**Comment by malte0811** [comment]
The crusher texture should be fixed in the latest Forge build. If I have time I'll try to figure out why the revolver is broken, I can't promise anything though.
